src/hotspot/share/services/heapDumper.cpp

```diff
@@ -1644,18 +1660,102 @@ class StickyClassDumper : public KlassClosure {
   }
 };
 
+// Large object heap dump support.
+// To avoid memory consumption, when dumping large objects such as huge array and
+// large objects whose size are larger than LARGE_OBJECT_DUMP_THRESHOLD, the scanned
+// partial object/array data will be send to backend directly instead of caching the
+// whole object/array internal buffer.
+// The HeapDumpLargeObjectList is used to save the large object when dumper scanning
+// the heap. The large objects could be added (push) parallelly by multiple dumpers.
+// But they will be removed (pop) serially only by the VM thread.
+class HeapDumpLargeObjectList : public CHeapObj<mtInternal> {
+ private:
+  class HeapDumpLargeObjectListElem : public CHeapObj<mtInternal> {
+   public:
+    HeapDumpLargeObjectListElem(oop obj) : _obj(obj), _next(NULL) { }
+    oop _obj;
+    HeapDumpLargeObjectListElem* _next;
+  };
+
+  HeapDumpLargeObjectListElem* _head;
+  uint _length;
+  uint _length1;
+
+ public:
+  HeapDumpLargeObjectList() : _head(NULL),  _length(0) { _length1 = 0; }
+
+  void atomic_push(oop obj) {
+    assert (obj != NULL, "sanity check");
+    HeapDumpLargeObjectListElem* entry = new HeapDumpLargeObjectListElem(obj);
+    if (entry == NULL) {
+      warning("Fail to allocate element for large object list");
+      return; 
+    }
+    assert (entry->_obj != NULL, "sanity check");
+    Atomic::inc(&_length1);
+    // HeapDumpLargeObjectListElem* head_val = _head;
+    while (true) {
+      HeapDumpLargeObjectListElem* old_head = Atomic::load_acquire(&_head);
+      HeapDumpLargeObjectListElem* new_head = entry;
+      if (Atomic::cmpxchg(&_head, old_head, new_head) == old_head) {
+        // successfully push
+        new_head->_next = old_head;
+        Atomic::inc(&_length);
+        assert(new_head->_obj == obj, "must equal");
+        return;
+      }
+    }
+  }
+
+  oop pop() {
+    if (_head == NULL) {
+      assert(_length == 0, "sanity check");
+      return NULL;
+    }
+    HeapDumpLargeObjectListElem* entry = _head;
+    if (_head->_next != NULL) {
+      _head = _head->_next;
+    }
+    entry->_next = NULL;
+    _length--;
+    assert (entry != NULL, "illegal larger object list entry");
+    oop ret = entry->_obj;
+    delete entry;
+    assert (ret != NULL, "illegal oop pointer");
+    return ret;
+  }
+
+  void drain(ObjectClosure* cl) {
+    printf("drain large object queue at length %d\n", _length);
+    while (_length > 0) {
+      cl->do_object(pop());
+    }
+  }
+
+  bool is_empty() {
+    return _length == 0;
+  }
+
+  uint length() { return _length; }
+
+  static const size_t LargeObjectSizeThreshold = 128 * (1 << 20); // 128 MB
```

Comment:
I think this threshold is much too large. I've instrumented the code to track the maximum memory use of the buffers and tested the max memory use and the performance for 2, 4 and 8 parallel iteration threads and for a threshold of 128 MB and 1 MB (using my scenario with a 32 GB heap):

128 MB threshold:
2 threads -> 57 MB max memory, 64.6 seconds duration
4 threads -> 88 MB max memory, 40.5 seconds duration
8 threads -> 170 MB max memory, 29.6 seconds duration

1MB threshold:
2 threads -> 8 MB max memory, 64.6 seconds duration
4 threads -> 15 MB max memory, 39.5 seconds duration
8 threads -> 32 MB max memory, 28.9 seconds duration

As you can see, using a much lower threshold has no discernable performance impact, but uses a lot less memory. Also note that the memory consumption can be much higher in the worst case scenario.

Reply:
I agree， Thanks for the evaluation!  I think 1MB is reasonable. Will made the change in next update.